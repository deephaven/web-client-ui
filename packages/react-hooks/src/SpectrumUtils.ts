// eslint-disable-next-line no-restricted-imports
import type { SpectrumTextFieldProps } from '@adobe/react-spectrum';
import { type KeyedItem } from '@deephaven/utils';
import type { DOMRefValue } from '@react-types/shared';

/**
 * Creates validation props for a Spectrum field. If `isValid` is true, returns
 * empty props. If false, returns  { errorMessage, validationState: 'invalid' }
 * @param isValid Whether props represent valid state
 * @param errorMessage Error message in the case `isValid` is false
 */
export function createValidationProps(
  isValid: boolean,
  errorMessage: string
): Pick<SpectrumTextFieldProps, 'validationState' | 'errorMessage'> {
  if (isValid) {
    return {};
  }

  return {
    errorMessage,
    validationState: 'invalid',
  };
}

/**
 * Extract DOM node from React Spectrum component ref.
 * @param ref
 */
export function extractSpectrumHTMLElement<
  THtml extends HTMLElement = HTMLElement,
>(ref: DOMRefValue<THtml> | null): HTMLElement | null {
  return ref?.UNSAFE_getDOMNode() ?? null;
}

/**
 * Extract lastElementChild from DOM element for given React Spectrum component
 * ref.
 * @param ref
 */
export function extractSpectrumLastChildHTMLElement<
  THtml extends HTMLElement = HTMLElement,
>(ref: DOMRefValue<THtml> | null): HTMLElement | null {
  const maybeHTMLElement = ref?.UNSAFE_getDOMNode().lastElementChild;
  return identityExtractHTMLElement(maybeHTMLElement);
}

/**
 * Find the popover associated with a given Spectrum ComboBox ref.
 * @param ref The ref to the Spectrum ComboBox component
 */
export function findSpectrumComboBoxScrollArea<
  THtml extends HTMLElement = HTMLElement,
>(ref: DOMRefValue<THtml> | null): HTMLElement | null {
  return findSpectrumPopoverScrollArea(ref, 'input');
}

/**
 * Find the popover associated with a given Spectrum Picker ref.
 * @param ref The ref to the Spectrum Picker component
 */
export function findSpectrumPickerScrollArea<
  THtml extends HTMLElement = HTMLElement,
>(ref: DOMRefValue<THtml> | null): HTMLElement | null {
  return findSpectrumPopoverScrollArea(ref, 'button');
}

/**
 * Find the popover associated with a given Spectrum component ref.
 * @param ref The ref to the Spectrum component
 * @param triggerElementType The type of element that triggers the popover
 */
export function findSpectrumPopoverScrollArea<
  K extends keyof HTMLElementTagNameMap,
  THtml extends HTMLElement = HTMLElement,
>(ref: DOMRefValue<THtml> | null, triggerElementType: K): HTMLElement | null {
  const maybeHTMLElement = ref?.UNSAFE_getDOMNode();
  const trigger = maybeHTMLElement?.querySelector(triggerElementType);
  const popupId = trigger?.getAttribute('aria-controls');

  const scrollArea = popupId == null ? null : document.getElementById(popupId);

  return scrollArea;
}

/**
 * Get the position of a selected item in a list of keyed items. The position is
 * based on the index, item height, and top offset.
 * @param keyedItems The list of keyed items
 * @param itemHeight The height of each item
 * @param selectedKey The key of the selected item
 * @param topOffset The offset from the top of the list (e.g. if there is top
 * padding surrounding the entire list)
 */
export async function getPositionOfSelectedItem<
  TKey extends string | number | boolean | undefined,
>({
  keyedItems,
  itemHeight,
  selectedKey,
  topOffset,
}: {
  keyedItems: KeyedItem<{ key?: TKey }, TKey>[];
  itemHeight: number;
  selectedKey: TKey | null | undefined;
  topOffset: number;
}): Promise<number> {
  const i = keyedItems.findIndex(
    item => (item.item?.key ?? item.key) === selectedKey
  );

  if (i <= 0) {
    return topOffset;
  }

  return itemHeight * i + topOffset;
}

/**
 * Returns the given object if it is an HTMLElement. Otherwise returns null.
 * @param maybeHTMLElement
 */
export function identityExtractHTMLElement<T>(
  maybeHTMLElement: T
): HTMLElement | null {
  return maybeHTMLElement instanceof HTMLElement ? maybeHTMLElement : null;
}
