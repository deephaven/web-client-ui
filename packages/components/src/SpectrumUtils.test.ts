import { theme } from '@react-spectrum/theme-default';
import { themeDHDefault } from './SpectrumUtils';

describe('themeDHDefault', () => {
  it('should merge Spectrum default with DH custom styles', () => {
    const { global, light, dark, medium, large } = theme;

    expect(themeDHDefault).toEqual({
      global,
      light: {
        ...light,
        'dh-spectrum-theme--light': 'mock.light',
      },
      dark: {
        ...dark,
        'dh-spectrum-theme--dark': 'mock.dark',
      },
      medium,
      large,
    });
  });
});
