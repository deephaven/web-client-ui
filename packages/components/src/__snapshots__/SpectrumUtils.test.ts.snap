// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`themeDHDefault should merge Spectrum default with DH custom styles 1`] = `
{
  "dark": {
    "spectrum--darkest": "spectrum--darkest_256eeb",
    "theme-spectrum-alias": "theme-spectrum-alias",
    "theme-spectrum-overrides": "theme-spectrum-overrides",
    "theme-spectrum-palette": "theme-spectrum-palette",
  },
  "global": {
    "spectrum": "spectrum_9e130c",
    "spectrum--dark": "spectrum--dark_9e130c",
    "spectrum--darkest": "spectrum--darkest_9e130c",
    "spectrum--large": "spectrum--large_9e130c",
    "spectrum--light": "spectrum--light_9e130c",
    "spectrum--lightest": "spectrum--lightest_9e130c",
    "spectrum--medium": "spectrum--medium_9e130c",
  },
  "large": {
    "spectrum--large": "spectrum--large_c40598",
  },
  "light": {
    "spectrum--light": "spectrum--light_a40724",
    "theme-spectrum-alias": "theme-spectrum-alias",
    "theme-spectrum-overrides": "theme-spectrum-overrides",
    "theme-spectrum-palette": "theme-spectrum-palette",
  },
  "medium": {
    "spectrum--medium": "spectrum--medium_4b172c",
  },
}
`;
